### PR TITLE
Fix change detection in CfgSimplifier::collapse_goto_chain

### LINKED ISSUE #75051: hang: issue-55608-captures-empty-region.rs with -Zmir-opt-level=2
### Code
`./src/test/ui/impl-trait/issues/issue-55608-captures-empty-region.rs`
```Rust
// This used to ICE because it creates an `impl Trait` that captures a
// hidden empty region.

// check-pass

fn server() -> impl FilterBase2 {
    segment2(|| { loop { } }).map2(|| "")
}

trait FilterBase2 {
    fn map2<F>(self, _fn: F) -> Map2<F> where Self: Sized { loop { } }
}

struct Map2<F> { _func: F }

impl<F> FilterBase2 for Map2<F> { }

fn segment2<F>(_fn: F) -> Map2<F> where F: Fn() -> Result<(), ()> {
    loop { }
}

fn main() { server(); }
```

### Meta
repo @ 12799ad60ca66b664699a0145f5ac5ccc399d4b8

### Error output

Rustc simply hangs when running
``build/x86_64-unknown-linux-gnu/stage1/bin/rustc ./src/test/ui/impl-trait/issues/issue-55608-captures-empty-region.rs  -Zmir-opt-level=2``

Edit: `perf top` while rustc is crunching on the hang:

````
  45,75%  librustc_driver-3c13214d88d90089.so  [.] rustc_mir::transform::simplify::CfgSimplifier::collapse_goto_chain
  28,35%  librustc_driver-3c13214d88d90089.so  [.] rustc_mir::transform::simplify::CfgSimplifier::simplify
   7,00%  [kernel]                             [k] mwait_idle_with_hints.constprop.0
   2,16%  librustc_driver-3c13214d88d90089.so  [.] rustc_middle::mir::terminator::TerminatorKind::successors_mut
   1,00%  librustc_driver-3c13214d88d90089.so  [.] rustc_middle::mir::terminator::Terminator::successors_mut
````

### LINKED ISSUE #75074: Cargo Build hangs forever with recent nighlies
With the following steps, I see the build hang indefinitely. This caused a 6 hour CI timeout for the nrf-rs organization. This does not happen with the stable/beta CI jobs, and did not happen previously with nightly. The following is a minimal repro of the CI steps of https://github.com/nrf-rs/nrf-hal/pull/189.

```bash
git clone https://github.com/kalkyl/nrf-hal.git
cd nrf-hal
git checkout comp
rustup override set nightly-2020-08-02-x86_64-unknown-linux-gnu
rustup target add thumbv7em-none-eabihf
cargo build --manifest-path examples/rtic-demo/Cargo.toml
```

### PR DESCRIPTION
**Pull request by tmiasko**
Check that the old target is different from the new collapsed one, before concluding that anything changed.

Fixes #75074
Fixes #75051

### CODE REVIEW

File: src/librustc_mir/transform/simplify.rs [hunk 1]
```diff
@@ -212,6 +210,7 @@ impl<'a, 'tcx> CfgSimplifier<'a, 'tcx> {
                 Terminator { kind: TerminatorKind::Goto { ref mut target }, .. } => target,
                 _ => unreachable!(),
             };
+            *changed |= *target != last;
```
Comment (oli-obk) [inline src/librustc_mir/transform/simplify.rs:213]: cc @ecstatic-morse We now have a test case for this!

File: src/test/ui/issues/issue-75704.rs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// Caused an infinite loop during SimlifyCfg MIR transform previously.
+//
+// build-pass
```
Comment (oli-obk) [inline src/test/ui/issues/issue-75704.rs:3]: does this really test this mir pass? don't you also need `-Zmir-opt-level=2`?

Comment (tmiasko) [reply src/test/ui/issues/issue-75704.rs:3]: For this test case no additional flags are necessary to reproduce the issue. Unlike the one from #75051 which also needed `-Zmir-opt-level=2`.

Comment (oli-obk) [reply src/test/ui/issues/issue-75704.rs:3]: ah, superb! thanks

**Pull request merged**